Make README reflect what multicodec currently is used for (#134)
multiformats/multicodec: README.md

[![](https://img.shields.io/badge/freenode-%23ipfs-blue.svg?style=flat-square)](https://webchat.freenode.net/?channels=%23ipfs)
[![](https://img.shields.io/badge/readme%20style-standard-brightgreen.svg?style=flat-square)](https://github.com/RichardLitt/standard-readme)

> > Save space by using the predefined multicodec table

Reviewer: This is a *claim*. I think the subtitle should reflect what multicodec *is*. I'd describe it as a canonical table enumerating well-known content types.

Author: What about something like "Canonical table of of codecs used by various multiformats". I still don't like that one, if anyone can come up with something better, that would be great.

## Table of Contents

- [Motivation](#motivation)
- [Description](#description)
- [Examples](#examples)
- [Multicodec table](#multicodec-table)
  - [Adding new multicodecs to the table](#adding-new-multicodecs-to-the-table)
- [Implementations](#implementations)
- [FAQ](#faq)
- [Contribute](#contribute)
- [License](#license)

## Motivation

> Multicodec is an agreed-upon codec table. It is designed for use in binary representations, such as keys or identifiers (i.e [CID](https://github.com/ipld/cid)).

Reviewer: What's the definition of a "codec"? We're using it in an unconventional way. Usually the codec is a device or software that performs encoding and decoding.

Reviewer 2: In the case of IPLD that’s exactly what codecs do, they encode and decode blocks.

Author: @Stebalien Do I recall correctly that you had a good answer to the question, why it is "codec" and not "content type". If yes, would you mind adding that as a FAQ?

Reviewer 2: content type has an existing meaning, and if you use that meaning then you could apply multiple content types to many codecs, and multiple codecs to a single content type, which means they cannot be synonymous.

Example: A block of `dag-json` can also be interpreted using `json` and the semantic meaning of the links would b stripped. The “content type” is still JSON. Similarly, you could interpret as `raw` and lose all semantic meaning. Content type is essentially what we’ve tried to describe as a “format” but unfortunately there’s a lot of older libraries and documentation that use the term format interchangeably with codec.

To add the confusion if we were to re-use content type, we already need to apply content types in UnixFS but we have to apply them to a “file” which is actually a dag connecting many blocks of raw binary data. Nowhere in that graph would you be able to use a content-type in place of the codec as a correct signal, it has to be added to the File metadata, or potentially wrapped in content-type metadata (the UnixFSv1 approach).

In short, a codec tells you how to decode block data into IPLD Data Model (JSON types + Links) and a content type tells an application how to interpret data (which may or may not involve parsing). There’s certainly some overlap, video file content types signal a literal codec for interpretation, but our interpretation is limited to how to interpret a single block into Data Model.

## Description

> The code of a multicodec is usually encoded as unsigned varint as defined by [multiformats/unsigned-varint](https://github.com/multiformats/unsigned-varint). It is then used as a prefix to identify the data that follows.

Reviewer: Let's remove usages from the description. Prefixing data with the multicodec is a valid use case, but not the only one. We can add a section on "Known/prescribed usages".

Author: @raulk: Do you mean removing the whole thing, or just the last sentence.

Author: @raulk What are other use cases?

## Examples

> Multicodec is used in various [Multiformats](https://github.com/multiformats/multiformats). In [Multihash](https://github.com/multiformats/multihash) it is used to identify the hashes, in the machine-readable [Multiaddr](https://github.com/multiformats/multiaddr) to identify the protocol.

Reviewer: ```suggestion
Multicodec is used in various [Multiformats](https://github.com/multiformats/multiformats). In [Multihash](https://github.com/multiformats/multihash) it is used to identify the hashes, in the machine-readable [Multiaddr](https://github.com/multiformats/multiaddr) to identify components such as IP addresses, domain names, identities, etc.
```

## Multicodec table

> As multicodecs are used across several multiformats, there's now a single table for all of them which can be fount at [table.csv](/table.csv) inside this repo. There's also a sortable [viewer](https://ipfs.io/ipfs/QmXec1jjwzxWJoNbxQF5KffL8q6hFXm9QwUGaa3wKGk6dT/#title=Multicodecs&src=https://raw.githubusercontent.com/multiformats/multicodec/master/table.csv).

Reviewer: We should drop language that refers to an elder time, as the reader won't be interested in the past but rather the present.

```suggestion
Find the canonical table of multicodecs at [table.csv](/table.csv). There's also a sortable [viewer](https://ipfs.io/ipfs/QmXec1jjwzxWJoNbxQF5KffL8q6hFXm9QwUGaa3wKGk6dT/#title=Multicodecs&src=https://raw.githubusercontent.com/multiformats/multicodec/master/table.csv).
```
